dist/tools/pyterm/pyterm

```diff
@@ -662,33 +662,43 @@ class SerCmd(cmd.Cmd):
         if self.set_dtr == 1 or self.set_dtr == 0:
             self.ser.setDTR(self.set_dtr)
 
+    def _handle_serial_exception(self):
+        self.logger.warning("Serial port disconnected, waiting to "
+                            "get reconnected...")
+        self.ser.close()
+        time.sleep(1)
+        if os.path.exists(self.port):
+            self.logger.warning("Try to reconnect to %s again..."
+                                % (self.port))
+            try:
+                self.serial_connect()
+                self.logger.info("Reconnected to serial port %s" % self.port)
+            except serial.SerialException:
+                pass
+
+    def _read_char(self):
+        # check if serial port can be accessed.
+        sr = codecs.getreader("UTF-8")(self.ser,
+                                       errors='replace')
+        return sr.read(1)
+
     def reader(self):
         """Serial or TCP reader.
         """
         output = ""
         crreceived = False
         nlreceived = False
+        c_already_read = False
         while (1):
-            # check if serial port can be accessed.
-            try:
-                sr = codecs.getreader("UTF-8")(self.ser,
-                                               errors='replace')
-                c = sr.read(1)
-            # try to re-open it with a timeout of 1s otherwise
-            except (serial.SerialException, ValueError):
-                self.logger.warning("Serial port disconnected, waiting to "
-                                    "get reconnected...")
-                self.ser.close()
-                time.sleep(1)
-                if os.path.exists(self.port):
-                    self.logger.warning("Try to reconnect to %s again..."
-                                        % (self.port))
-                    try:
-                        self.serial_connect()
-                        self.logger.info("Reconnected to serial port %s" % self.port)
-                    except serial.SerialException:
-                        pass
-                continue
+            # one of the if branches might have read c already
+            if not c_already_read:
+                try:
+                    c = self._read_char()
+                # try to re-open it with a timeout of 1s otherwise
+                except (serial.SerialException, ValueError):
+                    self._handle_serial_exception()
+                    continue
```

Comment:
I think this part should be its own commit, it's a different issue no?

Reply:
Well it's just some putting of functionalities into functions, that I later reuse for the fix. Can put them into their own commit though, if you insist.

Reply:
Done and thus force pushed.